core/src/test/java/org/elasticsearch/index/replication/IndexLevelReplicationTests.java

```diff
@@ -299,6 +306,65 @@ public void testRequestFailureReplication() throws Exception {
         }
     }
 
+    public void testTranslogDedupOperations() throws Exception {
+        try (ReplicationGroup shards = createGroup(2)) {
+            shards.startAll();
+            int initDocs = shards.indexDocs(randomInt(10));
+            List<IndexShard> replicas = shards.getReplicas();
+            IndexShard replica1 = replicas.get(0);
+            IndexShard replica2 = replicas.get(1);
+
+            logger.info("--> Isolate replica1");
+            IndexRequest indexDoc1 = new IndexRequest(index.getName(), "type", "d1").source("{}", XContentType.JSON);
+            BulkShardRequest replicationRequest = indexOnPrimary(indexDoc1, shards.getPrimary());
+            for (int i = 1; i < replicas.size(); i++) {
+                indexOnReplica(replicationRequest, replicas.get(i));
+            }
+
+            final Translog.Operation op1;
+            final List<Translog.Operation> initOperations = new ArrayList<>(initDocs);
+            try (Translog.Snapshot snapshot = replica2.getTranslog().newSnapshot()) {
+                assertThat(snapshot.totalOperations(), equalTo(initDocs + 1));
+                for (int i = 0; i < initDocs; i++) {
+                    Translog.Operation op = snapshot.next();
+                    assertThat(op, is(notNullValue()));
+                    initOperations.add(op);
+                }
+                op1 = snapshot.next();
+                assertThat(op1, notNullValue());
+                assertThat(snapshot.next(), nullValue());
+                assertThat(snapshot.overriddenOperations(), equalTo(0));
+            }
+
+            // Make sure that replica2 receives translog ops (eg. op2) from replica1 and overwrites its stale operation (op1).
+            logger.info("--> Promote replica1 as the primary");
+            shards.promoteReplicaToPrimary(replica1);
+            shards.index(new IndexRequest(index.getName(), "type", "d2").source("{}", XContentType.JSON));
+            final Translog.Operation op2;
+            try (Translog.Snapshot snapshot = replica2.getTranslog().newSnapshot()) {
+                assertThat(snapshot.totalOperations(), greaterThanOrEqualTo(initDocs + 2));
+                op2 = snapshot.next();
+                assertThat(op2.seqNo(), equalTo(op1.seqNo()));
+                assertThat(op2.primaryTerm(), greaterThan(op1.primaryTerm()));
+                assertThat("Remaining of snapshot should contain init operations", snapshot, containsOperationsInAnyOrder(initOperations));
+                assertThat(snapshot.overriddenOperations(), greaterThanOrEqualTo(1));
+            }
+
+            // Make sure that peer-recovery transfers all but non-overridden operations.
+            IndexShard replica3 = shards.addReplica();
+            logger.info("--> Promote replica2 as the primary");
+            shards.promoteReplicaToPrimary(replica2);
+            logger.info("--> Recover replica3 from replica2");
+            recoverReplica(replica3, replica2);
```

Comment:
can you also assert that the contents of the shards are the same - see ReplicationGroup#assertAllEqual

Reply:
Currently, we don't rollback replicas; replica2 has doc1 and doc2 while replica1 (eg. the promoted primary) does not have doc1.

Reply:
sure, but you can remove replica1 with a comment about it?

Reply:
@bleskes, This becomes more interesting to me :).

- Replica1 has `{doc1}`
- Replica2 has `{doc1, doc2}`
- Replica3 can have either `{doc2}` only if operation-based recovery from replica2, or `{doc1, doc2}` if file-based recovery from replica2.

Reply:
Funny you say this as I have a similar thought this morning while thinking about another issue. That other issue caused me to add a todo on the #10708 ticket to deal with it (see replica recover with rollback). I'm fine with leaving the test as is for now (with a comment explaining the situation and a todo to fix this when possible)

Reply:
I have added a TODO for this.